db/infinispan/src/main/java/org/commonjava/indy/infinispan/data/InfinispanStoreDataManager.java

```diff
@@ -144,4 +159,78 @@ protected ArtifactStore putArtifactStoreInternal( StoreKey storeKey, ArtifactSto
         return stores.put( storeKey, store );
     }
 
+    @Override
+    public Set<Group> affectedBy( final Collection<StoreKey> keys )
+    {
+        final Set<Group> groups = new HashSet<>();
+        for ( StoreKey key : keys )
+        {
+            Set<StoreKey> affected = affectedByStores.get( key );
+            if ( affected != null )
+            {
+                affected = affected.stream().filter( k -> k.getType() == group ).collect( Collectors.toSet() );
+                for ( StoreKey gKey : affected )
+                {
+                    ArtifactStore store = getArtifactStoreInternal( gKey );
+                    groups.add( (Group) store );
+                }
+            }
+        }
+        return groups;
+    }
+
+    @Override
+    protected void refreshAffectedBy( final ArtifactStore store, final ArtifactStore original )
+            throws IndyDataException
+    {
+        if ( store == null )
+        {
+            return;
+        }
+        if ( store instanceof Group )
+        {
+            final Set<StoreKey> updatedConstituents = new HashSet<>( ((Group)store).getConstituents() );
+            final Set<StoreKey> originalConstituents;
+            if ( original != null )
+            {
+                originalConstituents = new HashSet<>( ((Group)original).getConstituents() );
+            }
+            else
+            {
+                originalConstituents = new HashSet<>();
+            }
+
+            final Set<StoreKey> added = new HashSet<>();
+            final Set<StoreKey> removed = new HashSet<>();
+            for ( StoreKey updKey : updatedConstituents )
+            {
+                if ( !originalConstituents.contains( updKey ) )
+                {
+                    added.add( updKey );
+                }
+            }
+
+            for ( StoreKey oriKey : originalConstituents )
+            {
+                if ( !updatedConstituents.contains( oriKey ) )
+                {
+                    removed.add( oriKey );
+                }
+            }
+
+            final Set<StoreKey> allChangedSubs = new HashSet<>( added );
+            allChangedSubs.addAll( removed );
+
+            for ( StoreKey key : allChangedSubs )
+            {
+                final ArtifactStore sub = getArtifactStoreInternal( key );
+                refreshAffectedBy( sub, null );
+            }
+        }
+
+        final Set<Group> affectedBy = affectedByFromStores( Collections.singleton( store.getKey() ) );
```

Comment:
Hmm, so this looks like you might actually be denormalizing / flattening the affected-by relationship here before storing it? I wonder how sensitive this might be when you have:

* Group G contains Group H
* Group H contains RemoteRepository R
* Group G affected-by contains RemoteRepository R

Then: 

* Group H adds RemoteRepository S

Will Group G's affected-by cache also be updated to reflect its new affected-by relationship to RemoteRepository S?